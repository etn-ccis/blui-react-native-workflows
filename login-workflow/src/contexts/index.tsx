export * from './RegistrationWorkflowContext';
export * from './RegistrationContext';
export * from './SharedDictionaries';
export * from './ErrorContext';
export * from './AuthContext';
export * from './OktaAuthContext';