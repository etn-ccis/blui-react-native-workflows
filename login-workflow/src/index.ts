/**
 Copyright (c) 2021-present, Eaton

 All rights reserved.

 This code is licensed under the BSD-3 license found in the LICENSE file in the root directory of this source tree and at https://opensource.org/licenses/BSD-3-Clause.
 **/
/**
 * General docs
 *
 * @packageDocumentation
 * @module  @etn-sst/react-native-auth-ui
 * @preferred
 */
export { AuthNavigationContainer } from './screens/AuthNavigationContainer';
export type { AccountDetailInformation } from './subScreens/AccountDetails';
/** @ignore */
export { ChangePassword } from './screens/ChangePassword';
export { LoginHeaderSplash } from './components/LoginHeaderSplash';

export { DismissKeyboardView } from './components/DismissKeyboardView';
export { default as i18n } from './translations/i18n';

export * from '@brightlayer-ui/react-auth-shared';

declare global {
    // eslint-disable-next-line @typescript-eslint/no-namespace
    namespace ReactNativePaper {
        // eslint-disable-next-line @typescript-eslint/consistent-type-definitions
        interface ThemeColors {
            primaryPalette: {
                light: string;
                main: string;
                dark: string;
            };
            accentPalette: {
                light: string;
                main: string;
                dark: string;
            };
            errorPalette: {
                light: string;
                main: string;
                dark: string;
            };
            divider: string;
            textPalette: {
                primary: string;
                secondary: string;
                onPrimary: {
                    light: string;
                    main: string;
                    dark: string;
                };
                disabled: string;
            };
            actionPalette: {
                active: string;
                background: string;
                disabled: string;
                disabledBackground: string;
            };
            // overrides: $DeepPartial<ThemeOverrides>;
            // opacity: Partial<ThemeOpacity>;
        }
        // eslint-disable-next-line @typescript-eslint/consistent-type-definitions
        interface ThemeFonts {
            bold: ThemeFont;
        }
    }
}

export * from './new_architecture';
