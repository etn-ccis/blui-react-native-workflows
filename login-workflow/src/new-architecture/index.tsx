export * from './screens';
export * from './components';
