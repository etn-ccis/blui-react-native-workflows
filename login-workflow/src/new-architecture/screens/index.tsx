export * from './DemoScreen';
export * from './DemoRegistrationScreen';
export * from './Screen1';
export * from './Screen2';
export * from './EulaScreen';
export * from './DemoCreateAccountScreen';
