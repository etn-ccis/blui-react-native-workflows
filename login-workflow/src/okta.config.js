export default {
        clientId: process.env.OKTA_CLIENT_ID ?? '',
        redirectUri: process.env.OKTA_REDIRECT_URI ?? '',
        endSessionRedirectUri: process.env.OKTA_LOGOUT_REDIRECT_URI ?? '',
        discoveryUri: process.env.OKTA_ISSUER ?? '',
        scopes: ['openid', 'profile', 'offline_access', 'groups'],
        requireHardwareBackedKeyStore: false,
};