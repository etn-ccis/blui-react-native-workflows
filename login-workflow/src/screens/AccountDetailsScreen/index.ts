export * from './AccountDetailsScreenBase';
export * from './AccountDetailsScreen';
export * from './types';
