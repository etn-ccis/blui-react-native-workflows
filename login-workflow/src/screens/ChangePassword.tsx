/**
 * @packageDocumentation
 * @module Screens
 */

import React, { useCallback } from 'react';

// Components
import { View, StyleSheet, SafeAreaView, StatusBar, TextInput } from 'react-native';
import MatIcon from 'react-native-vector-icons/MaterialIcons';
import { PasswordRequirements } from '../components/PasswordRequirements';
import { TextInputSecure } from '../components/TextInputSecure';
import { Instruction } from '../components/Instruction';
import { Spinner } from '../components/Spinner';
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view';
import { ToggleButton } from '../components/ToggleButton';

// Styles
import { Body1, H6 } from '@brightlayer-ui/react-native-components';

// Hooks
import { ScrollView } from 'react-native-gesture-handler';
import { SimpleDialog } from '../components/SimpleDialog';
import { useTheme } from 'react-native-paper';

// Shared Auth Logic
import {
    // Actions
    initialTransitState,
    transitSuccess,
    transitStart,
    transitFailed,
    // Constants
    SPECIAL_CHAR_REGEX,
    LENGTH_REGEX,
    NUMBERS_REGEX,
    UPPER_CASE_REGEX,
    LOWER_CASE_REGEX,
    // Types
    PasswordRequirement,
    // Hooks
    useLanguageLocale,
    useInjectedUIContext,
} from '@brightlayer-ui/react-auth-shared';
import { CloseHeader } from '../components/CloseHeader';
import { WorkflowCardBody } from '..';

/**
 * @ignore
 */
const makeContainerStyles = (theme: ReactNativePaper.Theme): Record<string, any> =>
    StyleSheet.create({
        safeContainer: {
            height: '100%',
            backgroundColor: theme.colors.surface,
            marginBottom: 16,
            flex: 1,
            justifyContent: 'space-between',
        },
        scrollContainer: {
            flex: 1,
            alignContent: 'center',
        },
        scrollContentContainer: {
            alignSelf: 'center',
            maxWidth: 600,
        },
        mainContainer: {
            marginTop: 8,
            flex: 1,
            flexDirection: 'row',
            justifyContent: 'center',
            width: 'auto',
        },
        containerMargins: {
            marginHorizontal: 16,
        },
        containerSpacing: {
            marginVertical: 16,
        },
        iconContainer: {
            marginTop: 80,
            marginBottom: 32,
            alignSelf: 'center',
        },
    });

/**
 * @ignore
 */
const makeStyles = (theme: ReactNativePaper.Theme): Record<string, any> =>
    StyleSheet.create({
        inputMargin: {
            marginTop: 24,
        },
        sideBySideButtons: {
            flexDirection: 'row',
            justifyContent: 'space-evenly',
            alignItems: 'center',
            paddingVertical: 8,
        },
        headerText: {},
        bodyText: {
            color: theme.colors.text,
        },
        textSpacing: {
            marginVertical: 8,
        },
        wideButton: {
            height: 60,
            paddingVertical: 8,
        },
    });

/**
 * @param onChangePassword  Function to handle change password action.
 * @param onCancel  Function to handle cancel action.
 * @param onChangeComplete  Function to handle the on change complete action.
 * @param theme (Optional) react-native-paper theme partial to style the component.
 */
type ChangePasswordProps = {
    onChangePassword: (oldPassword: string, newPassword: string) => Promise<void>;
    onCancel: () => void;
    onChangeComplete: () => void;
    theme?: ReactNativePaper.Theme;
};

/**
 * Renders the content of the change password screen
 * (inputs current password, new password and confirm password).
 *
 * @category Component
 */
export const ChangePassword: React.FC<ChangePasswordProps> = (props) => {
    const theme = useTheme(props.theme);
    const [currentPasswordInput, setCurrentPasswordInput] = React.useState('');
    const [newPasswordInput, setNewPasswordInput] = React.useState('');
    const [confirmInput, setConfirmInput] = React.useState('');
    const [transitState, setTransitState] = React.useState(initialTransitState);
    const [hasAcknowledgedError, setHasAcknowledgedError] = React.useState(false);
    const { t } = useLanguageLocale();

    // styles
    const containerStyles = makeContainerStyles(theme);
    const styles = makeStyles(theme);

    // for continuing to the next input
    const newPasswordRef = React.useRef<TextInput>(null);
    const goToNewPasswordInput = (): void => newPasswordRef?.current?.focus();

    const confirmInputRef = React.useRef<TextInput>(null);
    const goToConfirmInput = (): void => confirmInputRef?.current?.focus();

    const defaultRequirements: PasswordRequirement[] = [
        {
            regex: LENGTH_REGEX,
            description: t('blui:PASSWORD_REQUIREMENTS.LENGTH'),
        },
        {
            regex: NUMBERS_REGEX,
            description: t('blui:PASSWORD_REQUIREMENTS.NUMBERS'),
        },
        {
            regex: UPPER_CASE_REGEX,
            description: t('blui:PASSWORD_REQUIREMENTS.UPPER'),
        },
        {
            regex: LOWER_CASE_REGEX,
            description: t('blui:PASSWORD_REQUIREMENTS.LOWER'),
        },
        {
            regex: SPECIAL_CHAR_REGEX,
            description: t('blui:PASSWORD_REQUIREMENTS.SPECIAL'),
        },
    ];
    const { passwordRequirements = defaultRequirements } = useInjectedUIContext();

    const areValidMatchingPasswords = useCallback((): boolean => {
        for (let i = 0; i < passwordRequirements.length; i++) {
            if (!new RegExp(passwordRequirements[i].regex).test(newPasswordInput)) return false;
        }
        return confirmInput === newPasswordInput;
    }, [passwordRequirements, newPasswordInput, confirmInput]);

    const spinner = transitState.transitInProgress ? <Spinner hasHeader={false} /> : <></>;

    const changePassword = async (): Promise<void> => {
        try {
            setHasAcknowledgedError(false);
            setTransitState(transitStart());
            await props.onChangePassword(currentPasswordInput, newPasswordInput);
            setTransitState(transitSuccess());
        } catch (error) {
            if (error instanceof Error) {
                setTransitState(transitFailed(error.message));
            } else throw error;
        }
    };

    const errorDialog = (
        <SimpleDialog
            title={t('blui:MESSAGES.ERROR')}
            bodyText={transitState.transitErrorMessage ?? ''}
            visible={transitState.transitErrorMessage !== null && !hasAcknowledgedError}
            onDismiss={(): void => {
                setHasAcknowledgedError(true);
            }}
        />
    );

    let statusBar: JSX.Element = <></>;
    statusBar = (
        <StatusBar
            backgroundColor={theme.colors?.primaryPalette?.main || theme.colors.primary}
            barStyle={'light-content'}
        />
    );

    return transitState.transitSuccess ? ( // if the password was changed
        <View
            style={{
                flex: 1,
                height: '100%',
                backgroundColor: theme.colors.surface,
                position: 'absolute',
                top: 0,
                width: '100%',
            }}
        >
            <CloseHeader title={'Change Password'} backAction={props.onChangeComplete} />
            <SafeAreaView style={[containerStyles.safeContainer, { flexGrow: 1 }]}>
                {statusBar}
                <View style={[containerStyles.mainContainer]}>
                    <ScrollView
                        style={containerStyles.scrollContainer}
                        contentContainerStyle={containerStyles.scrollContentContainer}
                    >
                        <MatIcon
                            name={'check'}
                            color={theme.colors.placeholder}
                            style={containerStyles.iconContainer}
                            size={100}
                        />
                        <View style={[containerStyles.containerMargins, containerStyles.containerSpacing]}>
                            <H6 style={[styles.headerText, styles.textSpacing]}>
                                {t('blui:CHANGE_PASSWORD.PASSWORD_CHANGED')}
                            </H6>
                            <Body1 style={[styles.bodyText, styles.textSpacing]}>
                                {t('blui:CHANGE_PASSWORD.SUCCESS_MESSAGE')}
                            </Body1>
                        </View>
                    </ScrollView>
                </View>
                <View style={[styles.wideButton, containerStyles.containerMargins]}>
                    <View style={{ flex: 1 }}>
                        <ToggleButton
                            text={t('blui:ACTIONS.LOG_IN')}
                            style={{ marginHorizontal: 16 }}
                            onPress={props.onChangeComplete}
                        />
                    </View>
                </View>
            </SafeAreaView>
        </View>
    ) : (
        // if the password hasn't been changed yet
        <View style={{ position: 'absolute', top: 0, bottom: 0 }}>
            <CloseHeader title={'Change Password'} backAction={props.onCancel} />
            <KeyboardAwareScrollView
                contentContainerStyle={{ flex: 1, height: '100%', backgroundColor: theme.colors.surface }}
            >
                <SafeAreaView style={[containerStyles.safeContainer, { flexGrow: 1 }]}>
                    {statusBar}
                    {spinner}
                    {errorDialog}
                    <View
                        style={{
                            flexDirection: 'row',
                            justifyContent: 'center',
                            width: 'auto',
                        }}
                    >
                        <View style={{ maxWidth: 600 }}>
                            <Instruction text={t('blui:CHANGE_PASSWORD.PASSWORD_INFO')} />
                        </View>
                    </View>
                    <ScrollView>
                        <WorkflowCardBody style={[containerStyles.containerMargins, containerStyles.mainContainer]}>
                            <View style={{ width: '100%', maxWidth: 600 }}>
                                <TextInputSecure
                                    label={t('blui:LABELS.CURRENT_PASSWORD')}
                                    value={currentPasswordInput}
                                    style={styles.inputMargin}
                                    autoCapitalize={'none'}
                                    returnKeyType={'next'}
                                    onChangeText={(text: string): void => setCurrentPasswordInput(text)}
                                    onSubmitEditing={(): void => {
                                        goToNewPasswordInput();
                                    }}
                                    blurOnSubmit={false}
                                />

                                <TextInputSecure
                                    label={t('blui:LABELS.NEW_PASSWORD')}
                                    ref={newPasswordRef}
                                    value={newPasswordInput}
                                    style={styles.inputMargin}
                                    autoCapitalize={'none'}
                                    returnKeyType={'next'}
                                    onChangeText={(text: string): void => setNewPasswordInput(text)}
                                    onSubmitEditing={(): void => {
                                        goToConfirmInput();
                                    }}
                                    blurOnSubmit={false}
                                />

                                <PasswordRequirements style={{ paddingTop: 8 }} passwordText={newPasswordInput} />

                                <TextInputSecure
                                    ref={confirmInputRef}
                                    label={t('blui:CHANGE_PASSWORD.CONFIRM_NEW_PASSWORD')}
                                    value={confirmInput}
                                    style={styles.inputMargin}
                                    autoCapitalize={'none'}
                                    returnKeyType={'done'}
                                    error={confirmInput !== '' && newPasswordInput !== confirmInput}
                                    onChangeText={(text: string): void => setConfirmInput(text)}
                                    onSubmitEditing={
                                        currentPasswordInput === '' || !areValidMatchingPasswords()
                                            ? undefined
                                            : changePassword
                                    }
                                />
                            </View>
                        </WorkflowCardBody>
                    </ScrollView>
                    <View style={[styles.sideBySideButtons, containerStyles.containerMargins]}>
                        <View style={{ flex: 1 }}>
                            <ToggleButton
                                text={t('blui:CHANGE_PASSWORD.UPDATE')}
                                disabled={currentPasswordInput === '' || !areValidMatchingPasswords()}
                                // eslint-disable-next-line @typescript-eslint/no-misused-promises
                                onPress={changePassword}
                            />
                        </View>
                    </View>
                </SafeAreaView>
            </KeyboardAwareScrollView>
        </View>
    );
};
