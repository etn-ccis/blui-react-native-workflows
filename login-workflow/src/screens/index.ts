export * from './DemoScreen';
export * from './RegistrationSuccessScreen';
export * from './EulaScreen';
export * from './CreateAccountScreen';
export * from './VerifyCodeScreen';
export * from './SuccessScreen';
export * from './AccountDetailsScreen';
export * from './CreatePasswordScreen';
export * from './ExistingAccountSuccessScreen';
export * from './ForgotPasswordScreen';
export * from './ChangePasswordScreen';
export * from './ContactScreen';
