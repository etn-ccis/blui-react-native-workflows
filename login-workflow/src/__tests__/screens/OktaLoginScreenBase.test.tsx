// OktaLoginScreenBase.test.tsx
import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import { OktaLoginScreenBase } from 'src/screens';
import { View } from 'react-native';

describe('OktaLoginScreenBase', () => {
    test('renders Okta Login Screen Base component', () => {
        render(<OktaLoginScreenBase />);
        expect(<OktaLoginScreenBase />).toMatchSnapshot()
    });

    test('renders self-register button if selfRegisterButtonLabel is provided', () => {
        const { getByTestId } = render(<OktaLoginScreenBase showSelfRegistration={true} selfRegisterButtonLabel="Register now!" />);
        const registerButton = getByTestId('blui-okta-login-self-register-label');
        expect(registerButton).toBeTruthy();
        expect(registerButton.props.children).toBe('Register now!');
    });

    test('renders contact support section if showContactSupport is true', () => {
        const { getByTestId } = render(<OktaLoginScreenBase showContactSupport={true} />);
        const contactSupportWrapper = getByTestId('blui-okta-login-contact-support-wrapper');
        expect(contactSupportWrapper).toBeTruthy();
    });

    test('renders contact support label with correct text', () => {
        const { getByTestId } = render(<OktaLoginScreenBase showContactSupport={true} contactSupportLabel="Contact Support" />);
        const contactSupportLabel = getByTestId('blui-okta-login-contact-support-label');
        expect(contactSupportLabel).toBeTruthy();
        expect(contactSupportLabel.props.children).toBe('Contact Support');
    });

    test('calls handleContactSupport when contact support label is pressed', () => {
        const handleContactSupport = jest.fn();
        const { getByTestId } = render(<OktaLoginScreenBase showContactSupport={true} onContactSupport={handleContactSupport} />);
        const contactSupportLabel = getByTestId('blui-okta-login-contact-support-label');
        fireEvent.press(contactSupportLabel);
        expect(handleContactSupport).toHaveBeenCalled();
    });

    test('renders footer content', () => {
        const { getByTestId } = render(<OktaLoginScreenBase footer={<View>Footer Content</View>} />);
        const footerContent = getByTestId('blui-okta-login-footer');
        expect(footerContent).toBeTruthy();
        expect(footerContent.props.children).toMatchSnapshot('Footer Content');
    });

    test('renders cybersecurity badge if showCyberSecurityBadge is true', () => {
        const { getByTestId } = render(<OktaLoginScreenBase showCyberSecurityBadge={true} />);
        const cyberSecurityBadgeWrapper = getByTestId('blui-okta-login-cyber-security-badge-wrapper');
        expect(cyberSecurityBadgeWrapper).toBeTruthy();
    });

    test('renders cybersecurity badge image', () => {
        const imgSource = require('../../assets/images/cybersecurity_certified.png');
        const { queryByTestId } = render(<OktaLoginScreenBase showCyberSecurityBadge={true} />);
        const cyberSecurityBadgeImage = queryByTestId('blui-okta-login-cyber-security-badge-image');
        expect(cyberSecurityBadgeImage).toBeTruthy();
        expect(queryByTestId('blui-okta-login-cyber-security-badge-image')?.props.source).toBe(imgSource);
    });
});