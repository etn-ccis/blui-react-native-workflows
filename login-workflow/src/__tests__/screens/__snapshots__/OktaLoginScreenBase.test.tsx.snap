// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OktaLoginScreenBase renders Okta Login Screen Base component 1`] = `<OktaLoginScreenBase />`;

exports[`OktaLoginScreenBase renders footer content: Footer Content 1`] = `
<View>
  Footer Content
</View>
`;

exports[`OktaLoginScreenBase renders self-register button if selfRegisterButtonLabel is provided 1`] = `null`;
